### PR TITLE
Re-enable local function region analysis

### PR DESCRIPTION
**Pull request by agocke**
Removes the code disabling region analysis for local functions and fixes
code that disabled capturing.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs [hunk 1]
```diff
@@ -5171,7 +5190,7 @@ static void Main()
         /*</bind>*/
     }
 }");
-            Assert.False(results.Succeeded);
+            Assert.True(results.Succeeded);
```
Comment (sharwell) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: ❓ Do we want to add assertions here for `results.Captured`?

Comment (agocke) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/RegionAnalysisTests.cs]: @sharwell This is going into a feature branch, I'm writing full tests up now. I just wanted to get the initial commit in.

**Comment by agocke** [comment]
test this please

**Comment by agocke** [comment]
test this please

**Comment by agocke** [comment]
ping @dotnet/roslyn-compiler for review

File: src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs [hunk 2]
```diff
@@ -1527,8 +1527,7 @@ protected void VisitStatementsWithLocalFunctions(BoundBlock block)
                 {
                     if (stmt.Kind == BoundKind.LocalFunctionStatement)
                     {
-                        VisitLocalFunctionStatement(
-                            (BoundLocalFunctionStatement)stmt);
+                        VisitAlways(stmt);
```
Comment (khyperia) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: Why was this changed? Same with accessibility? I assume just for style/code cleanup, but I just want to be sure.

Comment (agocke) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs]: Actually this is semantically important. VisitAlways marks the entry and exit for regions. When just visiting we skipped over the entry and exit, so statements were being assigned to the wrong regions.

**Comment by ivanbasov** [comment]
>         [Fact, WorkItem(21499, "https://github.com/dotnet/roslyn/issues/21499")]

Likely, you can remove this attribute here and below

---
Refers to: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs:4967 in f066cae. [](commit_id = f066cae0143a86613487a3646282d14137a41f2c, deletion_comment = False)

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 3]
```diff
@@ -6433,7 +6465,8 @@ void F()
 ";
             var edits = GetTopEdits(src1, src2);
 
-            edits.VerifySemanticDiagnostics();
+            edits.VerifySemanticDiagnostics(
+                Diagnostic(RudeEditKind.AccessingCapturedVariableInLambda, "f1", "this", "local function"));
```
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs]: >"local function" [](start = 89, length = 16)

CSharpFeaturesResources.local_function

File: src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs [hunk 4]
```diff
@@ -5850,7 +5864,8 @@ int f1(int a1)
             var edits = GetTopEdits(src1, src2);
 
             // y is no longer captured in f2
-            edits.VerifySemanticDiagnostics();
+            edits.VerifySemanticDiagnostics(
+                Diagnostic(RudeEditKind.NotAccessingCapturedVariableInLambda, "f2", "y", "local function"));
```
Comment (ivanbasov) [inline src/EditorFeatures/CSharpTest/EditAndContinue/StatementEditingTests.cs]: >"local function") [](start = 89, length = 17)

CSharpFeaturesResources.local_function
